File: 3757.number-of-effective-subsequences.go
#
# @lc app=leetcode id=3757 lang=golang
#
# [3757] Number of Effective Subsequences
#

# @lc code=start
func countEffective(nums []int) int {
	mod := int64(1000000007)

	full_or := 0
	for _, num := range nums {
		full_or |= num
	}

	bits := []int{}
	for b := 0; b < 32; b++ {
		if (full_or & (1 << b)) != 0 {
			bits = append(bits, b)
		}
	}

	m := len(bits)
	if m == 0 {
		return 0
	}

	max_mask := 1 << m

	F := make([]int64, max_mask)
	for _, num := range nums {
		w := 0
		for j := 0; j < m; j++ {
			if (num & (1 << bits[j])) != 0 {
				w |= (1 << j)
			}
		}
		F[w]++
	}

	sub_sum := make([]int64, max_mask)
	copy(sub_sum, F)
	for j := 0; j < m; j++ {
		bit := 1 << j
		for u := 0; u < max_mask; u++ {
			if (u & bit) != 0 {
				sub_sum[u] += sub_sum[u ^ bit]
			}
		}
	}

	ans := int64(0)

	all_mask := max_mask - 1

	var cntbits func(int) int = func(x int) int {
		c := 0
		x2 := x
		for x2 > 0 {
			c++
			x2 &= x2 - 1
		}
		return c
	}

	popc := make([]int, max_mask)
	for i := 0; i < max_mask; i++ {
		popc[i] = cntbits(i)
	}

	var powmod func(int64, int64, int64) int64 = func(base, exp, md int64) int64 {
		res := int64(1)
		base %= md
		for exp > 0 {
			if exp&1 == 1 {
				res = res * base % md
			}
			base = base * base % md
			exp >>= 1
		}
		return res
	}

	for t := 1; t <= all_mask; t++ {
		comp := all_mask ^ t
		zero_cnt := sub_sum[comp]

		p := popc[t]
		ways := powmod(2, zero_cnt, mod)

		if p%2 == 1 {
			ans = (ans + ways) % mod
		} else {
			ans = (ans - ways + mod) % mod
		}
	}

	return int(ans)
}
# @lc code=end

File: 3757.number-of-effective-subsequences.java
#
# @lc app=leetcode id=3757 lang=java
#
# [3757] Number of Effective Subsequences
#

# @lc code=start
class Solution {
    public int countEffective(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        int S = 0;
        for (int x : nums) S |= x;
        int k = 0;
        for (int b = 0; b < 32; b++) {
            if ((S & (1 << b)) != 0) k++;
        }
        int[] bits = new int[k];
        int idx = 0;
        for (int b = 0; b < 32; b++) {
            if ((S & (1 << b)) != 0) {
                bits[idx++] = b;
            }
        }
        int MS = 1 << k;
        long[] fr = new long[MS];
        for (int num : nums) {
            int mask = 0;
            for (int j = 0; j < k; j++) {
                if ((num & (1 << bits[j])) != 0) {
                    mask |= (1 << j);
                }
            }
            fr[mask]++;
        }
        long[] dp = new long[MS];
        for (int i = 0; i < MS; i++) dp[i] = fr[i];
        for (int b = 0; b < k; b++) {
            for (int mask = 0; mask < MS; mask++) {
                if ((mask & (1 << b)) == 0) {
                    dp[mask | (1 << b)] += dp[mask];
                }
            }
        }
        final int MOD = 1000000007;
        long[] pow2 = new long[n + 1];
        pow2[0] = 1;
        for (int i = 1; i <= n; i++) {
            pow2[i] = pow2[i - 1] * 2L % MOD;
        }
        long ans = 0;
        int full = (1 << k) - 1;
        for (int m = 1; m < MS; m++) {
            int pop = Integer.bitCount(m);
            long disj = dp[full ^ m];
            long ways = pow2[(int) disj];
            if ((pop & 1) == 1) {
                ans = (ans + ways) % MOD;
            } else {
                ans = (ans + MOD - ways) % MOD;
            }
        }
        return (int) ans;
    }
}
# @lc code=end